---
  title: "Runs of Homozygosity Analysis on a sheep dataset"
author: "Najla Abassi"
date: "`r Sys.Date()`"
output:
  html_document:
  theme: lumen
---
  
  ```{r setup, include=FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  eval = FALSE,
  warning = FALSE)
```

```{r load-lib}
library("xlsx")
library("factoextra")
library("FactoMineR")
library("BEDMatrix")
library("ggplot2")
library("gridExtra")
library("tidyverse")
# Data preprocessing and Quality Control

The original data was in the .bed .bim .fam format. The first step we did is to perform QC using the following metrics:
  
  * remove heterosomes
* SNPs with > 10% missing genotypes
* SNPs with MAF < 5%
* SNPs that are not in HW equilibrium

```{bash}
plink --bfile data --not-chr 23 24 25 26 --geno 0.1 --maf 0.05 --hwe 0.0001 --make-bed --out data_QC
```

We also need to make sure that we don't have duplicates in our data as the data comes from different sources. This step comes after merging all datasets. We remove all individuals with a PiHat = 1.

```{bash}
plink --bfile pop_merged --extract plink.prune.in --make-bed pop_pruned
plink --bfile pop_pruned --genome --min 0.05 --out ibd_report
plink --bfile pop_pruned --make-bed --out pruned_without_duplicates --remove duplucates.txt
```

# ROH identification

Now we can proceed with characterizing ROH in the processed dataset. 

```{r id_func}
# call all functions
source("./R/ROH_identification_param_functions.R")
source(".R/functions_statictical_analysis_using_plink_outputs.R")
source("./R/rohi_functions.R")
```


```{r}
hardy_path <- "/path/to/hardy_filter.hwe"
bim_path <- "path/to/data_QC.bim"
ni <- 100 # number of individuals as example

y <- homozyg_snp(hardy_path, ni)
x <- homozyg_density(bim_path)
round(y,0)
round(x,0)
```

Now we can identify ROH with the following command:

```{bash}
plink --bfile data_QC --homozyg-window-snp 50 --homozyg-window-het 1 --homozyg-window-missing 2 --homozyg-gap 100 --homozyg-kb 500 --homozyg-density x --homozyg-snp y --out homozyg_filter
```

where `x` is the minimum number of SNPs to have less than 5% of ROH generated by chance, and `y` is the SNP density.

## Statistical analysis

### Number of ROH and Total run length

```{r nb_length_roh}
indiv_path <- "/path/to/indiv_path.hom.indiv"

roh_summary_table <- roh_stat(indiv_path)
roh_summary_table

```

### FROH

```{r froh}
FROH_stat <- F_roh(indiv_path,bim_path)
FROH_stat
```

### FHOM

```{r fhom}
het_path <- "/path/to/het_filter.het"

FHOM_stat <- F_hom(het_path)
FHOM_stat

```

## Genomic distribution of ROH (ROH islands)

```{r rohi_detect}
hom <- read.table("/path/to/homozyg_filter.hom", header = TRUE)
populations <- c("list all population names as vector")
roh_islands_1 <- data.frame()
roh_islands_n <- data.frame() # we create for each population an empty dataframe; here we show the example for 2 populations

for (pop in populations) {
  POP <- hom[hom$FID == pop, ]
  roh_islands <- data.frame()
  for (chr in 1:22) {
    roh_islands_chr <- get_RHOi(POP, chr, pop)
    roh_islands <- rbind.data.frame(roh_islands, roh_islands_chr)
  }

  # Storing the results in separate variables based on the population
  if (pop == "1") {
    roh_islands_1 <- roh_islands
  } else if (pop == "n") {
    roh_islands_n <- roh_islands
  } 
}


head(roh_islands_1)
head(roh_islands_n)

```

Now we can save results as csv files

```{r save_results}
write.table(roh_islands_1,"rohi_1.csv",quote = F, col.names = T,row.names = F, sep = ",")

write.table(roh_islands_n,"rohi_n.csv",quote = F, col.names = T,row.names = F, sep = ",")

```

We saved the ROHi results in a directory called "rohi"

```{r summ_by_pop}
summary <- rohi_summ_pop("rohi")
summary

```

```{r save_summary}
write.table(summary,"summary_rohi.csv",quote = F, col.names = T,row.names = F, sep = ",")

```

## Functional Annotation

```{r prot_char}
library(biomaRt)
DATA1 <- read.csv("/path/to/rohi/rohi_1.csv")
DATA2 <- read.csv("/path/to/rohi/rohi_n.csv")
protein_table_1 <- get_Prot(DATA1)
protein_table_n <- get_Prot(DATA2)

head(protein_table_1)
head(protein_table_n)

```

```{r save_prot_results}
library("xlsx")
write.xlsx(protein_table_1,"protein_table_1.xls", sheetName = "Sheet1", col.names = T, row.names = F, append = F)

write.xlsx(protein_table_n,"protein_table_n.xls", sheetName = "Sheet1", col.names = T, row.names = F, append = F)

```
# NB: Plots, PCA, Cryptic relatedness will be added soon...

# Session Info {-}
```{r}
sessionInfo()
```


